### PR TITLE
[CARBONDATA-2206] support lucene index datamap

### PR DESCRIPTION
**Pull request by jackylk**
This PR is an initial effort to integrate lucene as an index datamap into carbondata. 
A new module called carbondata-lucene is added to support lucene datamap:
1. Add LuceneFineGrainDataMap, implement FineGrainDataMap interface.
2. Add LuceneCoarseGrainDataMap, implement CoarseGrainDataMap interface.
3. Support writing lucene index via LuceneDataMapWriter. 
4. Implement LuceneDataMapFactory 
5. A UDF called `TEXT_MATCH` is added, use it to do filtering on string column by lucene

User can use lucene datamap as:
```
    CREATE TABLE main(id INT, name STRING, city STRING, age INT)
    STORED BY 'carbondata'

    CREATE DATAMAP dm ON TABLE main
    USING 'org.apache.carbondata.datamap.lucene.LuceneFineGrainDataMapFactory'

    SELECT * FROM main WHERE TEXT_MATCH('name:bob')
```

This query will show all rows whose name is bob

 - [X] Any interfaces changed?
 No
 - [X] Any backward compatibility impacted?
 No
 - [X] Document update required?
No
 - [X] Testing done
Test will be added in subsequent PR
 - [X] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 
NA

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3933/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2688/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2691/

File: core/src/main/java/org/apache/carbondata/core/scan/filter/intf/ExpressionType.java [hunk 1]
```diff
@@ -42,5 +42,6 @@
   TRUE,
   STARTSWITH,
   ENDSWITH,
-  CONTAINSWITH
+  CONTAINSWITH,
+  MATCH
```
Comment (QiangCai) [inline core/src/main/java/org/apache/carbondata/core/scan/filter/intf/ExpressionType.java]: please use TEXT_MATCH

File: datamap/lucene/pom.xml [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+  <build>
+    <testSourceDirectory>src/test/scala</testSourceDirectory>
+    <resources>
+      <resource>
+        <directory>src/resources</directory>
+      </resource>
+      <resource>
+        <directory>.</directory>
+        <includes>
+          <include>CARBON_SPARK_INTERFACELogResource.properties</include>
```
Comment (QiangCai) [inline datamap/lucene/pom.xml]: not require

File: datamap/lucene/src/main/java/org/apache/carbondata/datamap/lucene/FilterExpressParser.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+import org.apache.lucene.analysis.Analyzer;
+import org.apache.lucene.analysis.standard.StandardAnalyzer;
+import org.apache.lucene.queryparser.classic.ParseException;
+import org.apache.lucene.queryparser.classic.QueryParser;
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.Query;
+
+@InterfaceAudience.Internal
+public class FilterExpressParser extends QueryParser {
```
Comment (QiangCai) [inline datamap/lucene/src/main/java/org/apache/carbondata/datamap/lucene/FilterExpressParser.java]: remove unuseful class

File: datamap/lucene/src/main/java/org/apache/carbondata/datamap/lucene/LuceneCoarseGrainDataMap.java [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+  public static final int ROWID_ID = 3;
+  /**
+   * searcher object for this datamap
+   */
+  private IndexSearcher indexSearcher = null;
+
+  /**
+   * default max values to return
+   */
+  private static int MAX_RESULT_NUMBER = 100;
```
Comment (QiangCai) [inline datamap/lucene/src/main/java/org/apache/carbondata/datamap/lucene/LuceneCoarseGrainDataMap.java]: better to config it

File: datamap/lucene/src/main/java/org/apache/carbondata/datamap/lucene/LuceneCoarseGrainDataMapFactory.java [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+  /**
+   * Get the datamap for segmentid
+   */
+  public List<CoarseGrainDataMap> getDataMaps(String segmentId) throws IOException {
+    List<CoarseGrainDataMap> lstDataMap = new ArrayList<>();
+    CoarseGrainDataMap dataMap = new LuceneCoarseGrainDataMap(analyzer);
+    try {
+      dataMap.init(new DataMapModel(
+          tableIdentifier.getTablePath() + "/Fact/Part0/Segment_" + segmentId + File.separator
```
Comment (QiangCai) [inline datamap/lucene/src/main/java/org/apache/carbondata/datamap/lucene/LuceneCoarseGrainDataMapFactory.java]: please invoke getSegmentDIr

File: core/src/main/java/org/apache/carbondata/core/scan/filter/intf/ExpressionType.java (hunk 1 above)
Comment (jackylk) [reply core/src/main/java/org/apache/carbondata/core/scan/filter/intf/ExpressionType.java]: fixed

File: datamap/lucene/pom.xml (hunk 2 above)
Comment (jackylk) [reply datamap/lucene/pom.xml]: fixed

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2692/

File: datamap/lucene/src/main/java/org/apache/carbondata/datamap/lucene/FilterExpressParser.java (hunk 3 above)
Comment (jackylk) [reply datamap/lucene/src/main/java/org/apache/carbondata/datamap/lucene/FilterExpressParser.java]: fixed

File: datamap/lucene/src/main/java/org/apache/carbondata/datamap/lucene/LuceneCoarseGrainDataMap.java (hunk 4 above)
Comment (jackylk) [reply datamap/lucene/src/main/java/org/apache/carbondata/datamap/lucene/LuceneCoarseGrainDataMap.java]: I will leave it to future PR

File: datamap/lucene/src/main/java/org/apache/carbondata/datamap/lucene/LuceneCoarseGrainDataMapFactory.java (hunk 5 above)
Comment (jackylk) [reply datamap/lucene/src/main/java/org/apache/carbondata/datamap/lucene/LuceneCoarseGrainDataMapFactory.java]: fixed

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3937/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2693/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2695/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3939/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2698/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3944/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3947/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2702/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3952/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2707/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2710/

**Comment by QiangCai** [comment]
LGTM

**Comment by jackylk** [comment]
merged into datamap branch

**Comment by ravipesala** [comment]
SDV Build Fail , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/3711/

**Pull request closed without merging**